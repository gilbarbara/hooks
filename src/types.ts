import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
