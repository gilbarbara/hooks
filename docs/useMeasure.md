# useMeasure
Get element measurements using the [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.

## Usage

```tsx
import React from 'react';
import { useMeasure } from '@gilbarbara/hooks';

function Component() {
  const ref = React.useRef<HTMLDivElement>(null);
  const dimensions = useMeasure(ref);

  return (
    <div ref={ref}>
      {JSON.stringify(dimensions, undefined, 2)}
    </div>
  );
}
```

## Reference

```typescript
interface UseMeasureResult extends Omit<DOMRectReadOnly, 'toJSON'> {
  absoluteHeight: number;
  absoluteWidth: number;
}

useMeasure<T extends Element>(
  target: RefObject<T> | T | null | string,
  debounce = 0,
): UseMeasureResult;
```

> This hook uses a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver), so if you want to support
legacy browsers, install the [resize-observer-polyfill](https://www.npmjs.com/package/resize-observer-polyfill) in your app.
