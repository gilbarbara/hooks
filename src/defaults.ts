import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
