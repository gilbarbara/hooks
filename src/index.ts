export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
