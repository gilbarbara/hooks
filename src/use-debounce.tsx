import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
