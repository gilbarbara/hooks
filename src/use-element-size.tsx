import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
