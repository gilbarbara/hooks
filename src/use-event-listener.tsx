import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
