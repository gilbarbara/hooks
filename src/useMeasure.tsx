import { useState } from 'react';

import { defaultElementDimensions } from './defaults';
import { Target } from './types';
import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
import { useResizeObserver } from './useResizeObserver';
import { canUseDOM, getElement } from './utils';

export interface UseMeasureResult extends Omit<DOMRectReadOnly, 'toJSON'> {
  absoluteHeight: number;
  absoluteWidth: number;
}

function parseFloatValue(value: string) {
  const parsed = parseFloat(value);

  return Number.isNaN(parsed) ? 0 : parsed;
}

function getElementMeasure(element: Element | null): UseMeasureResult {
  if (!canUseDOM() || !element) {
    return defaultElementDimensions;
  }

  const { bottom, height, left, right, top, width, x, y } = element.getBoundingClientRect();

  const {
    borderBottom,
    borderLeft,
    borderRight,
    borderTop,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
  } = getComputedStyle(element);

  return {
    absoluteHeight:
      height -
      parseFloatValue(paddingTop) -
      parseFloatValue(paddingBottom) -
      parseFloatValue(borderTop) -
      parseFloatValue(borderBottom),
    absoluteWidth:
      width -
      parseFloatValue(paddingLeft) -
      parseFloatValue(paddingRight) -
      parseFloatValue(borderLeft) -
      parseFloatValue(borderRight),
    bottom,
    height,
    left,
    right,
    top,
    width,
    x,
    y,
  };
}

export function useMeasure<T extends Element>(target: Target<T>, debounce = 0): UseMeasureResult {
  const [element, setElement] = useState(getElement(target));
  const [dimensions, setDimensions] = useState<UseMeasureResult>(getElementMeasure(element));

  const entry = useResizeObserver(element, debounce);

  useIsomorphicLayoutEffect(() => {
    const nextElement = getElement(target);

    setElement(nextElement);
    setDimensions(getElementMeasure(nextElement));
  }, [target]);

  useIsomorphicLayoutEffect(() => {
    if (!entry) {
      return;
    }

    const { bottom, height, left, right, top, width, x, y } = entry.contentRect;
    const { blockSize, inlineSize } = entry.borderBoxSize[0];

    setDimensions({
      absoluteHeight: blockSize,
      absoluteWidth: inlineSize,
      bottom,
      height,
      left,
      right,
      top,
      width,
      x,
      y,
    });
  }, [entry]);

  return dimensions;
}
