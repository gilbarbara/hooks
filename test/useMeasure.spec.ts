import { RefObject } from 'react';
import { act, renderHook } from '@testing-library/react';
import { mockResizeObserver } from 'jsdom-testing-mocks';

import {
  mockGetBoundingClientRectResponse,
  mockGetComputedStyleResponse,
  mockResizeObserveResponse,
} from './__fixtures__/data';

import { useMeasure, UseMeasureResult } from '../src/useMeasure';

const resizeObserver = mockResizeObserver();

function createElement(id = 'root') {
  const rootElement = document.createElement('div');

  rootElement.id = id;
  rootElement.innerHTML = 'Hello World';

  return rootElement;
}

const getComputedStyleData = mockGetComputedStyleResponse;

const { getComputedStyle } = window;

vi.spyOn(window, 'getComputedStyle').mockImplementation(el => ({
  ...getComputedStyle(el),
  ...getComputedStyleData,
}));

vi.spyOn(Element.prototype, 'getBoundingClientRect').mockImplementation(
  () => mockGetBoundingClientRectResponse,
);

describe('useMeasure', () => {
  const ref = { current: null } as RefObject<Element | null>;
  const rootElement = createElement();

  beforeAll(() => {
    document.body.appendChild(rootElement);
  });

  beforeEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    vi.restoreAllMocks();

    document.body.removeChild(rootElement);
  });

  it.each([{ target: 'element' }, { target: '#root' }, { target: 'ref' }])(
    'should return the dimensions with a "$target" target',
    ({ target }) => {
      let result: RefObject<UseMeasureResult> | null = null;
      let rerender: () => void;

      if (target === 'ref') {
        ref.current = rootElement;
        ({ rerender, result } = renderHook(() => useMeasure(ref.current)));
      } else if (target.startsWith('#')) {
        ({ rerender, result } = renderHook(() => useMeasure(target)));
      } else {
        ({ rerender, result } = renderHook(() => useMeasure(rootElement)));
      }

      expect(result?.current).toMatchSnapshot('initial');

      resizeObserver.mockElementSize(rootElement, mockResizeObserveResponse);

      act(() => {
        resizeObserver.resize();
      });

      rerender();
      expect(result.current).toMatchSnapshot('after resize');

      const { current } = result;

      rerender();
      expect(result.current).toBe(current);
    },
  );

  it('should re-initialize if target changes', async () => {
    const { rerender, result } = renderHook((target: string = '#app') => useMeasure(target));

    expect(result.current).toMatchSnapshot('initial');
    window.getComputedStyle = el => ({ ...getComputedStyle(el), ...getComputedStyleData });
    resizeObserver.mockElementSize(rootElement, mockResizeObserveResponse);

    rerender('#root');

    act(() => {
      resizeObserver.resize();
    });

    expect(result.current).toMatchSnapshot('after render');

    const { current } = result;

    rerender('#root');
    expect(result.current).toBe(current);
  });

  it('should return the default dimensions with an invalid string selector', () => {
    const { rerender, result } = renderHook(() => useMeasure('#app'));

    expect(result.current).toMatchSnapshot('initial');

    const { current } = result;

    rerender('#app');
    expect(result.current).toMatchSnapshot('after render');
    expect(result.current).toBe(current);
  });
});
