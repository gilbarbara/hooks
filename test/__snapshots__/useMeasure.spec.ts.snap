// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useMeasure > should re-initialize if target changes > after render 1`] = `
{
  "absoluteHeight": 100,
  "absoluteWidth": 640,
  "bottom": 100,
  "height": 100,
  "left": 0,
  "right": 640,
  "top": 0,
  "width": 640,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should re-initialize if target changes > initial 1`] = `
{
  "absoluteHeight": 0,
  "absoluteWidth": 0,
  "bottom": 0,
  "height": 0,
  "left": 0,
  "right": 0,
  "top": 0,
  "width": 0,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the default dimensions with an invalid string selector > after render 1`] = `
{
  "absoluteHeight": 0,
  "absoluteWidth": 0,
  "bottom": 0,
  "height": 0,
  "left": 0,
  "right": 0,
  "top": 0,
  "width": 0,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the default dimensions with an invalid string selector > initial 1`] = `
{
  "absoluteHeight": 0,
  "absoluteWidth": 0,
  "bottom": 0,
  "height": 0,
  "left": 0,
  "right": 0,
  "top": 0,
  "width": 0,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the dimensions with a "'#root'" target > after resize 1`] = `
{
  "absoluteHeight": 100,
  "absoluteWidth": 640,
  "bottom": 100,
  "height": 100,
  "left": 0,
  "right": 640,
  "top": 0,
  "width": 640,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the dimensions with a "'#root'" target > initial 1`] = `
{
  "absoluteHeight": 92,
  "absoluteWidth": 632,
  "bottom": 120,
  "height": 100,
  "left": 200,
  "right": 200,
  "top": 20,
  "width": 640,
  "x": 20,
  "y": 120,
}
`;

exports[`useMeasure > should return the dimensions with a "'element'" target > after resize 1`] = `
{
  "absoluteHeight": 100,
  "absoluteWidth": 640,
  "bottom": 100,
  "height": 100,
  "left": 0,
  "right": 640,
  "top": 0,
  "width": 640,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the dimensions with a "'element'" target > initial 1`] = `
{
  "absoluteHeight": 92,
  "absoluteWidth": 632,
  "bottom": 120,
  "height": 100,
  "left": 200,
  "right": 200,
  "top": 20,
  "width": 640,
  "x": 20,
  "y": 120,
}
`;

exports[`useMeasure > should return the dimensions with a "'ref'" target > after resize 1`] = `
{
  "absoluteHeight": 100,
  "absoluteWidth": 640,
  "bottom": 100,
  "height": 100,
  "left": 0,
  "right": 640,
  "top": 0,
  "width": 640,
  "x": 0,
  "y": 0,
}
`;

exports[`useMeasure > should return the dimensions with a "'ref'" target > initial 1`] = `
{
  "absoluteHeight": 92,
  "absoluteWidth": 632,
  "bottom": 120,
  "height": 100,
  "left": 200,
  "right": 200,
  "top": 20,
  "width": 640,
  "x": 20,
  "y": 120,
}
`;
